# 第四章。排除故障

在使用 Node.js 的过程中，您不可避免地要调试一些讨厌的 bug。所以，让我们提前期待他们，为那天做好计划。

# 测井

有一些方法我们可以用来调试我们的软件；我们要看的第一个是日志。记录消息最简单的方法就是使用`console`。在前面的大多数例子中`console`被用来描述正在发生的事情，而不需要看到整个 HTTP 请求和响应，从而使事情变得更加可读和简单。

一个例子是:

```js
var Http = require( 'http' );

Http.createServer( function( request, response ) {
    console.log( 
        'Received request', 
        request.method,
        request.url 
    )

    console.log( 'Returning 200' );

    response.writeHead( 200, { 'Content-Type': 'text/plain' } );
    response.end( 'Hello World\n' );

}).listen( 8000 );

console.log( 'Server running on port 8000' );
```

运行此示例将在控制台上记录请求和响应:

```js
[~/examples/example-6]$ node server.js
Server running on port 8000
Received request GET /
Returning 200
Received request GET /favicon.ico
Returning 200
Received request GET /test
Returning 200
Received request GET /favicon.ico
Returning 200

```

如果我们使用的是接受中间件的框架，比如 express，我们可以使用一个简单的`npm`包，叫做 **摩根**；您可以在[https://www.npmjs.com/package/morgan](https://www.npmjs.com/package/morgan)找到套餐:

```js
[~/examples/example-7]$ npm install morgan
[~/examples/example-7]$ npm install router

```

我们可以通过使用`require`将它带入到我们的代码中，并将其作为中间件添加来使用它:

```js
var Morgan = require( 'morgan' ),
    Router = require( 'router' ),
    Http = require( 'http' );

router = new Router( );

router.use( Morgan( 'tiny' ) ); 

/* Simple server */
Http.createServer( function( request, response ) {
    router( request, response, function( error ) {
        if( !error ) {
            response.writeHead( 404 );  
        } else {
            //Handle errors
            console.log( error.message, error.stack );
            response.writeHead( 400 );
        }
        response.end( '\n' );

    });
}).listen( 8000 );

console.log( 'Server running on port 8000' );

function getInfo ( request, response ) {
    var info = process.versions;

    info = JSON.stringify( info );
    response.writeHead( 200, { 'Content-Type': 'application/json' } );
    response.end( info );
}
router.get( '/info', getInfo );
```

当服务器运行时，我们可以看到每个请求和响应，而不必向每个处理程序添加日志:

```js
[~/examples/example-7]$ node server.js
Server running on port 8000
GET /test 404 - - 4.492 ms
GET /favicon.ico 404 - - 2.281 ms
GET /info 200 - - 1.120 ms
GET /info 200 - - 1.120 ms
GET /test 404 - - 0.199 ms
GET /info 200 - - 0.494 ms
GET /test 404 - - 0.162 ms
```

这种日志记录是查看服务器上正在使用什么以及每个请求需要多长时间的简单方法。在这里，您可以看到第一个请求花费的时间最长，然后它们变得更快。差别只有 3 毫秒；如果时间更长，这可能是一个大问题。

我们可以通过更改传递给摩根的格式来增加记录的信息，如图所示:

```js
router.use( Morgan( 'combined' ) );
```

通过运行服务器，您将看到更多信息，例如远程用户、请求的日期和时间、返回的内容量以及他们正在使用的客户端。

```js
[~/examples/example-7]$ node server.js 
Server running on port 8000
::1 - - [07/Jun/2015:11:09:03 +0000] "GET /info HTTP/1.1" 200 - "-" "--REMOVED---"

```

时机绝对是一个重要因素，因为它可以帮助你筛选堆积如山的原木。有些虫子就像一颗定时炸弹，等待着在周六凌晨 3 点爆炸。如果进程已经停止，日志已经消失，那么所有这些日志对我们来说毫无意义。还有另一个流行且有用的软件包叫做`bunyan,`，它将许多日志记录方法打包成一个。

Bunyan 带来了可写流写入日志的优势，无论是磁盘上的文件还是`stdout`。这允许我们为事后调试保存日志。您可以在[https://www.npmjs.com/package/bunyan](https://www.npmjs.com/package/bunyan)找到更多关于T1 的详细信息。

现在，让我们安装软件包。我们希望它在本地和全球安装，以便我们也可以将其用作命令行工具:

```js
 [~/examples/example-8]$ npm install –g bunyan
 [~/examples/example-8]$ npm install bunyan 

```

现在，让我们做一些记录:

```js
var Bunyan = require( 'bunyan' ),
    logger;

logger = Bunyan.createLogger( {
    name: 'example-8'
});
logger.info( 'Hello logging' );
```

运行我们的示例:

```js
[~/examples/example-8]$ node index.js
{"name":"example-8","hostname":"macbook.local","pid":2483,"level":30,"msg":"Hello logging","time":"2015-06-07T11:35:13.973Z","v":0}

```

这看起来不太好看，是吗？Bunyan 使用一个简单的结构化 JSON 字符串来保存消息；这使得解析、扩展和读取变得容易。Bunyan 附带了一个命令行界面实用程序，可以让一切变得美好。

如果我们使用该实用程序运行该示例，那么我们将看到输出格式良好:

```js
[~/examples/example-8]$ node index.js | bunyan
[2015-06-07T11:38:59.698Z]  INFO: example-8/2494 on macbook.local: Hello logging

```

如果我们再添加几个级别，您会在控制台上看到每个级别都有不同的颜色，以帮助我们识别它们:

```js
var Bunyan = require( 'bunyan' ),
    logger;
logger = Bunyan.createLogger( {
    name: 'example-8'
});
logger.trace( 'Trace' );
logger.debug( 'Debug' );
logger.info( 'Info' );
logger.warn( 'Warn' );
logger.error( 'Error' );
logger.fatal( 'Fatal' );

logger.fatal( 'We got a fatal, lets exit' );
process.exit( 1 );
```

让我们运行这个例子:

```js
[~/examples/example-8]$ node index.js | bunyan
[2015-06-07T11:39:55.801Z]  INFO: example-8/2512 on macbook.local: Info
[2015-06-07T11:39:55.811Z]  WARN: example-8/2512 on macbook.local: Warn
[2015-06-07T11:39:55.814Z] ERROR: example-8/2512 on macbook.local: Error
[2015-06-07T11:39:55.814Z] FATAL: example-8/2512 on macbook.local: Fatal
[2015-06-07T11:39:55.814Z] FATAL: example-8/2512 on macbook.local: We got a fatal, lets exit

```

如果您注意到，控制台上没有输出跟踪和调试。这是因为他们习惯于跟随节目的流程而不是关键信息，并且通常非常嘈杂。

我们可以通过在创建记录器时将此作为选项传递来更改我们希望看到的日志级别:

```js
logger = Bunyan.createLogger( {
    name: 'example-8',
    level: Bunyan.TRACE 
});
```

现在，当我们运行这个例子时:

```js
[~/examples/example-8]$ node index.js | bunyan
[2015-06-07T11:55:40.175Z] TRACE: example-8/2621 on macbook.local: Trace
[2015-06-07T11:55:40.177Z] DEBUG: example-8/2621 on macbook.local: Debug
[2015-06-07T11:55:40.178Z]  INFO: example-8/2621 on macbook.local: Info
[2015-06-07T11:55:40.178Z]  WARN: example-8/2621 on macbook.local: Warn
[2015-06-07T11:55:40.178Z] ERROR: example-8/2621 on macbook.local: Error
[2015-06-07T11:55:40.178Z] FATAL: example-8/2621 on macbook.local: Fatal
[2015-06-07T11:55:40.178Z] FATAL: example-8/2621 on macbook.local: We got a fatal, lets exit

```

我们通常不希望看到低于信息级别的日志，因为任何对事后调试有用的信息都应该使用信息或更高的级别来记录。

Bunyan 的 api 非常适合记录错误和对象的功能。它将正确的结构保存在其准备显示的 JSON 输出中:

```js
try {
    ref.go( );
} catch ( error ) {
    logger.error( error );
}
```

让我们运行示例:

```js
[~/examples/example-9]$ node index.js | bunyan
[2015-06-07T12:00:38.700Z] ERROR: example-9/2635 on macbook.local: ref is not defined
 ReferenceError: ref is not defined
 at Object.<anonymous> (~/examples/example-8/index.js:9:2)
 at Module._compile (module.js:460:26)
 at Object.Module._extensions..js (module.js:478:10)
 at Module.load (module.js:355:32)
 at Function.Module._load (module.js:310:12)
 at Function.Module.runMain (module.js:501:10)
 at startup (node.js:129:16)
 at node.js:814:3

```

如果我们查看示例并漂亮地打印它，我们会看到他们将其保存为错误:

```js
[~/examples/example-9]$ npm install -g prettyjson
[~/examples/example-9]$ node index.js | prettyjson
name:     example-9
hostname: macbook.local
pid:      2650
level:    50
err: 
 message: ref is not defined
 name:    ReferenceError
 stack: 
 """
 ReferenceError: ref is not defined
 at Object.<anonymous> (~/examples/example-8/index.js:9:2)
 at Module._compile (module.js:460:26)
 at Object.Module._extensions..js (module.js:478:10)
 at Module.load (module.js:355:32)
 at Function.Module._load (module.js:310:12)
 at Function.Module.runMain (module.js:501:10)
 at startup (node.js:129:16)
 at node.js:814:3
 """
msg:      ref is not defined
time:     2015-06-07T12:02:33.875Z
v:        0

```

这是很有用的，因为如果你只是记录一个错误，如果你使用`JSON.stringify`你会得到一个空的对象，或者如果你使用`toString`你会得到一条消息:

```js
try {
    ref.go( );
} catch ( error ) {
    console.log( JSON.stringify( error ) );
    console.log( error );
    console.log( {
        message: error.message
        name: error.name
        stack: error.stack
    });
}
```

让我们运行这个例子:

```js
[~/examples/example-10]$ node index.js
{}
[ReferenceError: ref is not defined]
{ message: 'ref is not defined',
 name: 'ReferenceError',
 stack: '--REMOVED--' }

```

使用`logger.error( error )`要比使用`logger.error( { message: error.message /*, ... */ } );`简单干净得多。

正如前面提到的，`bunyan`使用了流的概念，这意味着我们可以写入一个文件，`stdout`，或者我们希望扩展到的任何其他服务。

要写入文件，我们需要做的就是将其添加到设置时传递给`bunyan`的选项中:

```js
var Bunyan = require( 'bunyan' ),
    logger;

logger = Bunyan.createLogger( {
    name: 'example-11',
    streams: [
        {
            level: Bunyan.INFO,
            path: './log.log'   
        }
    ]
});
logger.info( process.versions );
logger.info( 'Application started' );
```

通过运行该示例，您将不会看到任何日志输出到控制台，但它们将被写入文件:

```js
 [~/examples/example-11]$ node index.js

```

如果您列出目录中的内容，您将看到一个新文件已经创建:

```js
[~/examples/example-11]$ ls 
index.js     log.log      node_modules

```

如果您阅读文件中的内容，您会发现日志已经被写入:

```js
[~/examples/example-11]$ cat log.log
{"name":"example-11","hostname":"macbook.local","pid":3614,"level":30,"http_parser":"2.3","node":"0.12.2","v8":"3.28.73","uv":"1.4.2-node1","zlib":"1.2.8","modules":"14","openssl":"1.0.1m","msg":"","time":"2015-06-07T12:29:46.606Z","v":0}
{"name":"example-11","hostname":"macbook.local","pid":3614,"level":30,"msg":"Application started","time":"2015-06-07T12:29:46.608Z","v":0}

```

我们可以运行这个到`bunyan`，以便很好地打印出来:

```js
[~/examples/example-11]$ cat log.log | bunyan
[~/examples/example-11]$ cat log.log | bunyan
[2015-06-07T12:29:46.606Z]  INFO: example-11/3614 on macbook.local:  (http_parser=2.3, node=0.12.2, v8=3.28.73, uv=1.4.2-node1, zlib=1.2.8, modules=14, openssl=1.0.1m)
[2015-06-07T12:29:46.608Z]  INFO: example-11/3614 on macbook.local: Application started

```

既然我们可以登录到一个文件，我们也希望能够看到显示的消息。如果我们只是登录一个文件，我们可以使用:

```js
[~/examples/example-11]$ tail -f log.log | bunyan

```

这将作为正在写入的文件记录到`stdout`；或者，我们可以向`bunyan`添加另一个流:

```js
logger = Bunyan.createLogger( {
    name: 'example-11',
    streams: [
        {
            level: Bunyan.INFO,
            path: './log.log'   
        },
        {
            level: Bunyan.INFO,
            stream: process.stdout
        }
    ]
});
```

运行该示例将向控制台显示日志:

```js
[~/examples/example-11]$ node index.js | bunyan
 [2015-06-07T12:37:19.857Z] INFO: example-11/3695 on macbook.local: (http_parser=2.3, node=0.12.2, v8=3.28.73, uv=1.4.2-node1, zlib=1.2.8, modules=14, openssl=1.0.1m) [2015-06-07T12:37:19.860Z] INFO: example-11/3695 on macbook.local: Application started

```

我们还可以看到日志已经追加到文件中:

```js
[~/examples/example-11]$ cat log.log | bunyan
 [2015-06-07T12:29:46.606Z]  INFO: example-11/3614 on macbook.local:  (http_parser=2.3, node=0.12.2, v8=3.28.73, uv=1.4.2-node1, zlib=1.2.8, modules=14, openssl=1.0.1m)
[2015-06-07T12:29:46.608Z]  INFO: example-11/3614 on macbook.local: Application started
[2015-06-07T12:37:19.857Z]  INFO: example-11/3695 on macbook.local:  (http_parser=2.3, node=0.12.2, v8=3.28.73, uv=1.4.2-node1, zlib=1.2.8, modules=14, openssl=1.0.1m)
[2015-06-07T12:37:19.860Z]  INFO: example-11/3695 on macbook.local: Application started

```

太好了，现在我们已经把日志记录下来了，我们该怎么办？

嗯，知道我们的错误发生在哪里是有帮助的，当你在这个地方有很多匿名函数的时候，它就开始变得非常混乱。如果您注意到在涵盖 HTTP 服务器的示例中，大多数函数都被命名了。当涉及回调时，这对于跟踪错误非常有帮助。

让我们看看这个例子:

```js
try {
    a = function( callback ) {
        return function( ) {
            callback( );
        };
    };
    b = function( callback ) {
        return function( ) {
            callback( );
        }
    };
    c = function( callback ) {
        return function( ) {
            throw new Error( "I'm just messing with you" ); 
        };
    };
    a( b( c( ) ) )( );
} catch ( error ) {
    logger.error( error );
}
```

可能看起来有点乱，那是因为确实如此。让我们运行以下示例:

```js
[~/examples/example-12]$ node index.js | bunyan
 [2015-06-07T12:51:11.665Z] ERROR: example-12/4158 on macbook.local: I'm just messing with you
 Error: I'm just messing with you
 at /Users/fabian/examples/example-12/index.js:19:10
 at /Users/fabian/examples/example-12/index.js:14:4
 at /Users/fabian/examples/example-12/index.js:9:4
 at Object.<anonymous> (/Users/fabian/examples/example-12/index.js:22:16)
 at Module._compile (module.js:460:26)
 at Object.Module._extensions..js (module.js:478:10)
 at Module.load (module.js:355:32)
 at Function.Module._load (module.js:310:12)
 at Function.Module.runMain (module.js:501:10)
 at startup (node.js:129:16)

```

您可以看到在我们的代码中没有函数名，并且与前几个函数不同，在堆栈跟踪中也没有命名。在 Node.js 中，函数的命名要么来自变量名，要么来自实际的函数名。例如，如果您使用`Cls.prototype.func`，那么名称将是`Cls.func`，但是如果您使用功能`func`，那么名称将是`func`。

您可以看到这里有一点好处，但是一旦您开始使用涉及`async`回调的模式，这就变得非常有用:

```js
[~/examples/example-13]$ npm install q

```

让我们在回调中抛出一个错误:

```js
var Q = require( 'q' );

Q( )
.then( function() {
    // Promised returned from another function
    return Q( )
    .then( function( ) {
        throw new Error( 'Hello errors' ); 
    });
})
.fail( function( error ) {
    logger.error( error );
});
```

运行我们的示例可以让我们:

```js
[~/examples/example-13]$ node index.js | bunyan
 [2015-06-07T13:03:57.047Z] ERROR: example-13/4598 on macbook.local: Hello errors
 Error: Hello errors
 at /Users/fabian/examples/example-13/index.js:12:9
 at _fulfilled (/Users/fabian/examples/example-13/node_modules/q/q.js:834:54)

```

这是开始难读的地方；为函数指定简单的名称可以帮助我们找到错误的来源:

```js
return Q( )
    .then( function resultFromOtherFunction( ) {
        throw new Error( 'Hello errors' ); 
    });
```

运行示例:

```js
[~/examples/example-13]$ node index.js | bunyan
 [2015-06-07T13:04:45.598Z] ERROR: example-13/4614 on macbook.local: Hello errors
 Error: Hello errors
 at resultFromOtherFunction (/Users/fabian/examples/example-13/index.js:12:9)
 at _fulfilled (/Users/fabian/examples/example-13/node_modules/q/q.js:834:54)

```

# 错误处理

调试的另一个方面是提前处理和预期错误。有三种方法可以处理我们的错误:

*   一个简单的`try` / `catch`
*   在流程层面抓住他们
*   捕捉域级别的错误

如果我们预期会发生错误，并且我们能够在不知道正在执行的任何操作的结果的情况下继续，或者我们可以处理并返回错误，那么`try` / `catch`函数就足够了，如图所示:

```js
function parseJSONAndUse( input ) {
    var json = null;
    try {
        json = JSON.parse( input );
    } catch ( error ) {
        return Q.reject( new Error( "Couldn't parse JSON" ) );
    }
    return Q( use( json ) );
}
```

捕捉错误的另一个简单方法是向您的进程添加一个错误处理程序；在这个级别上捕捉到的任何错误通常都是致命的，应该这样对待。应该有一个退出过程，您应该使用一个包，如`forever`或`pm2`:

```js
process.on( 'uncaughtException', function errorProcessHandler( error ) {
    logger.fatal( error );
    logger.fatal( 'Fatal error encountered, exiting now' );
    process.exit( 1 );
});
```

您应该总是在出现未发现的错误后退出进程。它未被捕获这一事实意味着您的应用处于未知状态，任何事情都可能发生。例如，您的 HTTP 路由器可能出现了错误，无法将更多请求路由到正确的处理程序。您可以在[https://nodejs . org/API/process . html # process _ event _ uncut exception](https://nodejs.org/api/process.html#process_event_uncaughtexception)上了解更多。

在全球范围内处理错误的更好方法是使用`domain`。有了域，你几乎可以把*沙箱*一组异步代码放在一起。

让我们从向服务器发出请求的角度来思考。我们发出请求，从数据库中读取，调用外部服务，写回数据库，做一些日志记录，做一些业务逻辑，我们期望完美的数据来自代码周围的外部来源。然而，在现实世界中并不总是这样，我们不能处理可能发生的每一个错误；此外，我们不想因为一个非常具体的请求出现一个错误就关闭整个服务器。这就是我们需要域名的地方。

让我们看看下面的例子:

```js
var Domain = require( 'domain' ),
    domain;

domain = Domain.create( );

domain.on( 'error', function( error ) {
    console.log( 'Domain error', error.message );
});

domain.run( function( ) {
    // Run code inside domain
    console.log( process.domain === domain );
    throw new Error( 'Error happened' ); 
});
```

让我们运行代码:

```js
[~/examples/example-14]$ node index.js
true
Domain error Error happened

```

此代码有问题；然而，当我们同步运行时，我们仍然将进程置于中断状态。这是因为错误冒泡到节点本身，然后被传递到活动域。

当我们在异步回调中创建域时，我们可以确定这个过程可以继续。我们可以用`process.nextTick`来模拟:

```js
process.nextTick( function( ) {
    domain.run( function( ) {
        throw new Error( 'Error happened' );
    });
    console.log( "I won't execute" );
}); 

process.nextTick( function( ) {
    console.log( 'Next tick happend!' );
});

console.log( 'I happened before everything else' );
```

运行该示例应显示正确的日志:

```js
[~/examples/example-15]$ node index.js
I happened before everything else
Domain error Error happened
Next tick happend!

```

# 总结

在本章中，我们已经介绍了一些事后调试方法来帮助我们发现错误，包括日志记录、命名实践和充分的错误处理。

在下一章中，我们将介绍应用的配置。

Prepared for Bentham Chang, Safari ID bentham@gmail.com User number: 2843974 © 2015 Safari Books Online, LLC. This download file is made available for personal use only and is subject to the Terms of Service. Any other use requires prior written consent from the copyright owner. Unauthorized use, reproduction and/or distribution are strictly prohibited and violate applicable laws. All rights reserved.